TASK-54 Add advanced exception support

KirSanich opened the pull request:
Добавление кастомного исключения к сервису, а также реализация глобального обработчика исключений

Commit:
TASK-54 Add advanced exception support

Добавление кастомного исключения к сервису, а также реализация глобального обработчика исключений

KirSanich commented on core/build.gradle.kts:
```diff
@@ -26,15 +26,19 @@ repositories {
 
 dependencies {
+
+    val springDocVersion = "1.6.9"
```
Небольше минорные правки, не относятся к самой логике этого пулл реквеста, но пусть уж будут тут.

KirSanich commented on core/src/main/kotlin/ru/surf/core/service/impl/EventServiceImpl.kt:
```diff
@@ -34,12 +35,14 @@ class EventServiceImpl(
 
     override fun getEvent(id: UUID): FullResponseEventDto {
-        val eventFromDb = eventRepository.findByIdOrNull(id) ?: throw NoSuchElementException()
+        val eventFromDb = eventRepository.findByIdOrNull(id)
+            ?: throw CoreServiceException.CODE.CS_DATABASE_ERROR.awaken("NO EVENT WITH ID $id")
```
Вот эта логика дублируется в 3 методах сразу, можно сделать запрос в репозитории и туда это вынести и вызывать уже 3 раза метод а-ля find..orThrowException, но не знаю насколько это обоснованно будет.

NikitaYakuntsev commented on core/src/main/kotlin/ru/surf/core/exception/CoreServiceException.kt:
```diff
@@ -0,0 +1,33 @@
+
+    enum class CODE(val description: String) {
+
+        CS_DATABASE_ERROR("Core service database error");
```
а когда случился конфликт по уникальному констреинту - это тоже будет дб эррор?
а если в базе табличка пропала - тоже бд эррор?

NikitaYakuntsev commented on core/src/main/kotlin/ru/surf/core/service/impl/EventServiceImpl.kt:
```diff
@@ -54,6 +57,10 @@ class EventServiceImpl(
 
-    override fun deleteEvent(id: UUID) = eventRepository.deleteById(id)
+    override fun deleteEvent(id: UUID) {
+        val eventFromDb = eventRepository.findByIdOrNull(id)
```
а что если тебе надо будет поменять сообщение об отсутствии эвента в базе? везде менять придется?

NikitaYakuntsev commented on core/src/main/kotlin/ru/surf/core/controller/CoreGlobalExceptionHandler.kt:
```diff
@@ -0,0 +1,21 @@
+    @ExceptionHandler(CoreServiceException::class)
+    fun databaseException(exception: CoreServiceException): ResponseEntity<ErrorDto> {
+        val errorDto = ErrorDto(exception.code.toString(),exception.message ?: "", LocalDateTime.now())
+        return ResponseEntity(errorDto, HttpStatus.BAD_REQUEST)
```
мне кстати нравится маппить сообщения об отсутствующих сущностях на 404 код

Commit:
TASK-54 Add-advanced-exception(refactor)

Разбиение одного исключения на несколько путем создания иерархии.

Commit:
TASK-54 Add-advanced-exception

Финальные правки по исключению, вынесения вызова и валидации метода DAL в отдельный метод в EventServiceImpl.kt

soup1ck commented on core/src/main/kotlin/ru/surf/core/exception/ExceptionType.kt:
```diff
@@ -0,0 +1,9 @@
+    SERVICE_EXCEPTION,
+    DATABASE_EXCEPTION,
+    CLIENT_EXCEPTION
+
```
по идее не должно быть в енаме строк пустых

soup1ck commented on core/src/main/kotlin/ru/surf/core/exception/ExceptionType.kt:
```diff
@@ -0,0 +1,9 @@
+package ru.surf.core.exception
+
+enum class ExceptionType {
+
```
тут тоже
